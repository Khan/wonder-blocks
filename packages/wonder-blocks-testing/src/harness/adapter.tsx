import * as React from "react";

import type {TestHarnessAdapter} from "./types";

type Props<TConfig, TAdapter extends TestHarnessAdapter<TConfig>> = {
    children: React.ReactNode;
    adapter: TAdapter;
    config: TConfig | null | undefined;
};

/**
 * Component that optionally renders a given adapter with the given config.
 *
 * If the config is nullish, then the children are rendered in a fragment,
 * otherwise the children are rendered within the given adapter with the
 * given config.
 */
export const Adapter = <TConfig, TAdapter extends TestHarnessAdapter<TConfig>>({
    children,
    adapter,
    config,
}: Props<TConfig, TAdapter>): React.ReactElement => {
    if (config == null) {
        return <>{children}</>;
    }
    return adapter(children, config);
};
