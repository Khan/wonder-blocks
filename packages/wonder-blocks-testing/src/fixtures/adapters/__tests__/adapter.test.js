// @flow
import {Adapter} from "../adapter.js";

import * as AdapterGroupModule from "../adapter-group.js";

jest.mock("../adapter-group.js");

describe("Adapter", () => {
    describe("#constructor", () => {
        it.each([null, 8, "", "     "])(
            "should throw if the name is invalid (%s)",
            (invalidName) => {
                // Arrange

                // Act
                const act = () => new Adapter(invalidName, () => {});

                // Assert
                expect(act).toThrowErrorMatchingSnapshot();
            },
        );

        it.each([null, 5])(
            "should throw if closeGroupFn is invalid",
            (invalidCloseGroupFn) => {
                // Arrange

                // Act
                const act = () => new Adapter("name", invalidCloseGroupFn);

                // Assert
                expect(act).toThrowErrorMatchingSnapshot();
            },
        );
    });

    describe("@name", () => {
        it("should match the value passed during construction", () => {
            // Arrange
            const name = "adapter_name";
            const adapter = new Adapter(name, () => {});

            // Act
            const result = adapter.name;

            // Assert
            expect(result).toBe(name);
        });
    });

    describe("#declareGroup", () => {
        it("should pass the closeGroupFn and options to the adapter group", () => {
            // Arrange
            const closeGroupFn = jest.fn();
            const adapter = new Adapter("adapter_name", closeGroupFn);
            const options = {
                title: "group_title",
                description: "group_description",
                getDefaultTitle: () => {
                    throw new Error("NOT IMPLEMENTED");
                },
            };
            const adapterGroupSpy = jest
                .spyOn(AdapterGroupModule, "AdapterGroup")
                .mockImplementation(() => {});

            // Act
            adapter.declareGroup(options);

            // Assert
            expect(adapterGroupSpy).toHaveBeenCalledWith(closeGroupFn, options);
        });

        it("should return the created AdapterGroup", () => {
            // Arrange
            const adapter = new Adapter("adapter_name", () => {});
            const adapterGroup: any = {
                whatAmI: "ADAPTER_GROUP",
            };
            jest.spyOn(AdapterGroupModule, "AdapterGroup").mockImplementation(
                () => adapterGroup,
            );

            // Act
            const result = adapter.declareGroup({
                title: "group_title",
                description: "group_description",
                getDefaultTitle: () => {
                    throw new Error("NOT IMPLEMENTED");
                },
            });

            // Assert
            expect(result).toBe(adapterGroup);
        });
    });
});
