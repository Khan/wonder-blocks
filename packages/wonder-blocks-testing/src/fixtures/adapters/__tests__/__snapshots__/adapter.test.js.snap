// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Adapter #constructor should throw if closeGroupFn is invalid 1`] = `"closeGroupFn must be a function"`;

exports[`Adapter #constructor should throw if closeGroupFn is invalid 2`] = `"closeGroupFn must be a function"`;

exports[`Adapter #constructor should throw if the name is invalid (     ) 1`] = `"name must be a non-empty string"`;

exports[`Adapter #constructor should throw if the name is invalid () 1`] = `"name must be a non-empty string"`;

exports[`Adapter #constructor should throw if the name is invalid (8) 1`] = `"name must be a string"`;

exports[`Adapter #constructor should throw if the name is invalid (null) 1`] = `"name must be a string"`;
